File: test/cpu/not.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"
#include "test.hpp"

#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
BOOST_FIXTURE_TEST_SUITE(NOT, Prepare)
struct F {
    F() {
        // NOT.B %D2
        TEST::SET_W(0, 0043000 | 2);
        TEST::SET_W(2, TEST_BREAK);

        // NOT.W %D2
        TEST::SET_W(4, 0043100 | 2);
        TEST::SET_W(6, TEST_BREAK);

        // NOT.L %D2
        TEST::SET_W(8, 0043200 | 2);
        TEST::SET_W(10, TEST_BREAK);

        jit_compile(0, 12);
    }
};
BOOST_AUTO_TEST_SUITE(R, *boost::unit_test::fixture<F>())
BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = 0x87;
    run_test(0);
    BOOST_TEST(cpu.D[2] == 0x78);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 1;
    run_test(0);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0xff;
    run_test(0);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = 0x8421;
    run_test(4);
    BOOST_TEST(cpu.D[2] == 0x7BDE);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 1;
    run_test(4);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0xffff;
    run_test(4);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = 0xF2345678;
    run_test(8);
    BOOST_TEST(cpu.D[2] == 0x0DCBA987);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 1;
    run_test(8);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0xffffffff;
    run_test(8);
    BOOST_TEST(cpu.Z);
}


BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/not.lua
-- NOT.B %D2
MEM[0].W = OCT "0043000" | 2
MEM[2].W = TEST.BREAK

-- NOT.W %D3
MEM[4].W = OCT "0043100" | 3
MEM[6].W = TEST.BREAK

-- NOT.L %D4
MEM[8].W = OCT "0043200" | 4
MEM[10].W = TEST.BREAK

JIT_COMPILE(0, 12)

local run = {}

run.B = {
    value = function()
        CPU.D[3].B = 0x87
        TEST.RUN(0)
        TEST.CHECK("CPU.D[3].B", 0x78)
        TEST.CHECK("CPU.CC.V", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    n = farray(TF, function(b)
        CPU.D[3].B = cond(b, 0x20, 0xC0)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.N", b)
    end),
    z = farray(TF, function(b)
        CPU.D[3].B = cond(b, 0xFF, 0)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", b)
    end)
}

run.W = {
    value = function()
        CPU.D[4].W = 0x8421
        TEST.RUN(4)
        TEST.CHECK("CPU.D[4].W", 0x7BDE)
        TEST.CHECK("CPU.CC.V", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    n = farray(TF, function(b)
        CPU.D[4].W = cond(b, 0x2000, 0xC000)
        TEST.RUN(4)
        TEST.CHECK("CPU.CC.N", b)
    end),
    z = farray(TF, function(b)
        CPU.D[4].W = cond(b, 0xFFFF, 0)
        TEST.RUN(4)
        TEST.CHECK("CPU.CC.Z", b)
    end)
}

run.L = {
    value = function()
        CPU.D[5].L = 0xF2345678
        TEST.RUN(8)
        TEST.CHECK("CPU.D[5].L", 0x0DCBA987)
        TEST.CHECK("CPU.CC.V", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    n = farray(TF, function(b)
        CPU.D[5].L = cond(b, 0x20000000, 0xC0000000)
        TEST.RUN(8)
        TEST.CHECK("CPU.CC.N", b)
    end),
    z = farray(TF, function(b)
        CPU.D[5].L = cond(b, 0xFFFFFFFF, 0)
        TEST.RUN(8)
        TEST.CHECK("CPU.CC.Z", b)
    end)
}

return run
